src/python/pants/engine/unions_test.py

```diff
@@ -19,3 +23,112 @@ class B:
+    # @runtime_checkable).
+
+    class Base(UnionBase):
+        def frobnicate(self) -> bool:
+            ...
+
+    # This will be the line to go with, using Protocol directly. The runtime_checkable decorator
+    # could be made optional, especially if using attributes, which it is not compatible with.. I
+    # was hoping for something cleaner, but maybe not too bad, considering the hackish workarounds
+    # required to make it work otherwise.
+    @runtime_checkable
+    class Base2(Protocol):
+        def some(self) -> str:
+            ...
+
+    class Base3(Protocol):
+        pass
+
+    class Base4(Protocol):
+        attr: int
+
+    # All bases of a Protocol must be protocols.. so we tripped up here, with the UnionBase.
+    class Derived(Base, Protocol):  # type: ignore[misc]
+        def gonk(self) -> None:
+            ...
+
+    class Derived2(Base2, Protocol):
+        def other(self) -> float:
+            ...
+
+    class A:
+        def frobnicate(self) -> bool:
+            return False
+
+    class B:
+        def some(self) -> str:
+            return "thing"
+
+        def frobnicate(self) -> bool:
+            return True
+
+    class C:
+        def some(self) -> str:
+            return "thing"
+
+    class D(A, C):
+        def gonk(self) -> None:
+            return None
+
+        def other(self) -> float:
+            return 1.2
+
+    union_membership = UnionMembership.from_rules(
+        [
+            UnionRule(Base, A),
+            UnionRule(Base, B),
+            UnionRule(Base2, B),
+            UnionRule(Base2, C),
+            UnionRule(Derived, D),
+            UnionRule(Derived2, D),
+        ]
+    )
+    assert union_membership == UnionMembership(
+        {
+            Base: FrozenOrderedSet([A, B]),
+            Base2: FrozenOrderedSet([B, C]),
+            Derived: FrozenOrderedSet([D]),
+            Derived2: FrozenOrderedSet([D]),
+        }
+    )
+
+    assert union_membership.has_members(Base)
+    assert not union_membership.has_members(UnionBase)
+    assert union_membership.is_member(Base, A())
+    assert not union_membership.is_member(Base2, A())
+    with pytest.raises(TypeError):
+        union_membership.is_member(UnionBase, A())
+
+    # wanted typecheck to err on the following, but didn't manage that, so settled to have a runtime
+    # check instead.
+    with pytest.raises(ValueError):
+        # Not OK - C does not fulfill the Base protocol
+        UnionRule(Base, C)
+
+    with pytest.raises(ValueError):
+        # Not OK - Base3 is not runtime checkable
+        UnionRule(Base3, C)
+
+    # Type check tests, with UnionBase vs clean Protocol classes
+
+    # mypy does not pick up our trick to turn Base into a Protocol, so this is not OK
+    a: Base = A()  # type: ignore[assignment]
+
+    # Base2 and Base3 are both explicitly Protocols, so this is OK
+    b: Base2 = B()
+    c: Base3 = C()
```

Comment:
> However, we do not get the mypy-level of type safety when registering Union members with UnionRule solely using Protocols, that was my caveat at the end, with the foo and bar assignment examples.

Assuming that the lines I'm commenting on here are actually correctly typechecked, then this seems like a good approach to me: IMO, if we're able to completely remove the "union" concept in favor of extending `Protocol`, and then renaming `UnionRule` to `ProtocolInstance` (strawname, but you get the idea), then that's great.

Reply:
These lines are typechecked correctly, so that's a win. However, I fail to incorporate this syntax into something to include in the list of rules.

I've posted a question on SO for this: https://stackoverflow.com/questions/68822297/python-protocol-implementation-class-type-check

Aahh.. unless you suggest that we use a few more lines while registering, which would accommodate mypy to do its thing..?

```python
class MyProto (Protocol):
  a: str

class Member:
  a: str

def rules():
  m: MyProto = Member()
  return [ProtocolInstance(MyProto, m), ...]]
```

I guess we could work with that.. maybe even be able to hide the assignment to be done generically in a dataclass.. reducing boilerplate.

Reply:
Got it. https://stackoverflow.com/questions/68822297/protocol-implementation-class-type-check clarifies the issue, thanks.

Reply:
I managed to find a way around it, finally :D 
more details on the SO answer I posted, but it's workable, and I think we can still translate that into a `UnionRule` as a first step, if we want, to make this less impact (and also be backwards compatible so we can use both in a transitional period)

Reply:
OK, so there seems to be a bug with how mypy treats Protocol class types. python/mypy#10988

Reply:
Circling back to renaming `UnionRule`.. I think we should be conservative with renaming too much, unless we "have" to. The concept of a union rule is well documented. Hopefully, it will continue to work referring to them as "unions" even if implemented using protocols, if we can keep the two compatible enough.

Reply:
Thanks for being cautious about renaming! I think we can actually do this fairly risk-free by aliasing UnionRule and @union and having a deprecation period. Or we can just stick with the term "union" as you suggest. Either way, no need to decide right now... :)

Reply:
I'm not particularly attached to the term `union` (especially since it has a different meaning in Python typing), so in the medium term (post deprecation) actually switching to `Protocol` makes sense to me. It gives us better alignment with Python typing (only with the addition of registration).